### ISSUE TITLE
Flux ignores kustomization.yaml

### ISSUE THREAD

**Issue by ar-qun** [fluxcd/flux2#3204]
### Describe the bug

After recently deploying a new cluster with GKE version 1.22 I receive the error below:

```
Kustomization/flux-system/${environment} dry-run failed, reason: Invalid, error: Kustomization.kustomize.toolkit.fluxcd.io "${environment}" is invalid: metadata.name: Invalid value: "${environment}": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

It seems that that kustomization.yaml file is somehow completely ignored. Because I compared contents of all the patch targets and they are clearly not patched.

When, I assume, trying to deploy this:
```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: ${environment}
  namespace: flux-system
spec:
  prune: True
  interval: 1m
  dependsOn:
    - name: namespaces
    - ....
  path: environments/${environment}/application
  sourceRef:
    kind: GitRepository
    name: flux-system
    namespace: flux-system
  postBuild:
    substitute:
      environment: ${environment}
```

### Steps to reproduce

1. Kubernetes version 1.22.12-gke.1200 - I am not sure about this step but that it is the only significant change at that stage

kustomization.yaml:
```yaml
# This manifest was generated by Terraform. DO NOT EDIT.
# Modify this file through the flux module
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- gotk-sync.yaml
- gotk-components.yaml
patches:
  - target:
      version: v1beta2
      group: kustomize.toolkit.fluxcd.io
      kind: Kustomization
      name: flux-system
      namespace: flux-system
    patch: |-
      - op: add
        path: /spec/postBuild
        value:
          substitute:
            environment: "dev"
```

### Expected behavior

Flux should pickup kustomization.yaml and apply all the patches in it.

### OS / Distro

WSL2,  Kubernetes version 1.22.12-gke.1200

### Flux version

v0.31.5, v0.35.0

### Flux check

flux check
► checking prerequisites
✗ flux 0.31.5 <0.35.0 (new version is available, please upgrade)
W1013 11:21:43.198854   10069 gcp.go:120] WARNING: the gcp auth plugin is deprecated in v1.22+, unavailable in v1.25+; use gcloud instead.
To learn more, consult https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke
✔ Kubernetes 1.22.12-gke.1200 >=1.20.6-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.25.0
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.26.0
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.22.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.29.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.27.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.30.0
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta1
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ gitrepositories.source.toolkit.fluxcd.io/v1beta2
✔ helmcharts.source.toolkit.fluxcd.io/v1beta2
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta2
✔ imagepolicies.image.toolkit.fluxcd.io/v1beta1
✔ imagerepositories.image.toolkit.fluxcd.io/v1beta1
✔ imageupdateautomations.image.toolkit.fluxcd.io/v1beta1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1beta2
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta1
✔ receivers.notification.toolkit.fluxcd.io/v1beta1
✔ all checks passed

### Git provider

GitHub

### Container Registry provider

ghcr.io

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ar-qun** [comment]
So I was able to solve the issue by renaming a kustomize.toolkit.fluxcd.io/v1beta2 from "${environment}" to dev. Weirdly enough after that the patches got applied everywhere and now I can rename it back to "${environment}".

This changes the scope of this issue so I will open a new one.

**Comment by stefanprodan** [comment]
Are you applying this with Terraform or the `flux bootstrap`? Only the Flux CLI supports customising the root files.

**Comment by joshuadmatthews** [comment]
I am seeing the same issue; except I am not Kustomizing the root. I am in AKS using Bicep to deploy my initial Kustomization and I noticed today that when I try to include the K8s Kustomization needed for Image Automation ACR credential cron job auth method either it doesn't deploy, or it does and the other files in my folder don't get deployed.

I had to move this k8s kustomiztion into its own top-level flux kustomization away from other manifests to get things working.

**Comment by stefanprodan** [comment]
> Can't seem to treat a kustomization.yaml like any other normal yaml file

Because is not a normal file, it's a Kustomize config not a Kubernetes object.

**Comment by ar-qun** [comment]
> Are you applying this with Terraform or with `flux bootstrap`? Only the Flux CLI supports customising the root files.

I am applying this with terraform. I modify the root files with with `flux-system/kustomization.yaml`.

**Comment by ar-qun** [comment]
So what seems to happen is that kustomize-controller loop during the initial bootstrap process behaves this way, which makes it behave unexpectedly in my specific use case:

1. First the kustomize.toolkit.fluxcd.io/v1beta2 is applied
2. kustomize-controller then loads the files specified over the path, which in my case is `path: environments/test/application`. Fixed in the description.
3. The files are picked up from the path but only the names are being read.
4. In my case the kustomize-controller **fails** because a file has variable substitution in `.metada.name`.
5. If the previous step wouldn't fail then the `kustomize.config.k8s.io/v1beta1` would be applied now, specifically the variable substitution.
6. Now the files that were picked up at step 3 are actually applied.

My conclusion would be that if the step 5 would be switched with step 4 it would solve the issue.

Could you confirm that is what is happening?

**Comment by stefanprodan** [comment]
@ar-qun no kustomize-controller does't handle the bootstrap, it's the Flux CLI that does it. Use `flux bootstrap` and it will work, since the CLI builds the `flux-system` overlay before it does the first apply.

**Comment by kingdonb** [comment]
We're trying to make heads or tails of this thread at Bug Scrub today and see whether there is any action needed.

Was the advice to try `flux bootstrap` helpful? Near as we can tell from the thread, the install may have been done with `kubectl apply -f flux-system/` and that would explain why the `kustomization.yaml` was skipped.

**Comment by joshuadmatthews** [comment]
I think the last comment basically explains my issue. Kustomization.yaml is either generated automatically to include all file in the directory or it defines the files explicitly. So by adding it in with other files that it didn’t define I was overriding the automatic version with a manual version that didn’t include the other files.

So in my case, simply placing it in a separate top level kustomization resolved the issue. Nothing to do with how I bootstrapped flux I don’t think.

**Comment by kingdonb** [comment]
Alright, thank you for the feedback 🎉 

Closing

**Comment by ar-qun** [comment]
> @ar-qun no kustomize-controller does't handle the bootstrap, it's the Flux CLI that does it. Use `flux bootstrap` and it will work, since the CLI builds the `flux-system` overlay before it does the first apply.

@stefanprodan Thank you for your explanation. Probably, I used the word "bootstrap" incorrectly. Let me rephrase my issue:

The `kustomization-controller` returns the error below after the installation of flux if the variable substitution is set in the name of a `Kustomization.kustomize.toolkit.fluxcd.io` file because the patches from `flux-system/kustomization.yaml` are not applied. If I remove the variable substitution, then the patches are applied and I can put the variable substitution back in the name.

```
Kustomization/flux-system/${environment} dry-run failed, reason: Invalid, error: Kustomization.kustomize.toolkit.fluxcd.io "${environment}" is invalid: metadata.name: Invalid value: "${environment}": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

**Comment by ar-qun** [comment]
> Alright, thank you for the feedback 🎉
> 
> Closing

Hi @kingdonb sadly my issue is as the creator of this case is still not resolved. Would it be possible to re-open it? I would gladly prepare a MRE for the next Bug Scrub? Would that work?

**Comment by stefanprodan** [comment]
The `flux-system` Kustomization was not designed to work like this. Substitutions should happen in other Kustomizations, the `flux-system` one is generated by `flux bootstrap` and it shouldn't be modified.

**Comment by ar-qun** [comment]
> The `flux-system` Kustomization was not designed to work like this. Substitutions should happen in other Kustomizations, the `flux-system` one is generated by `flux bootstrap` and it shouldn't be modified.

I am following the "Automate image updates to Git" guide that tells me that it is the way to do the patching.  Section "Using Native GCP GCR Auto-Login" specifically. Link below.

 https://fluxcd.io/flux/guides/image-update/#using-native-gcp-gcr-auto-login

**Comment by ar-qun** [comment]
Also apart from this issue with the variable substitution it works perfectly.

**Comment by stefanprodan** [comment]
> I am following the "Automate image updates to Git" guide that tells me that it is the way to do the patching.

There is no variable substitution used there. As I said, `flux-system` can't do substitutions.

**Comment by ar-qun** [comment]
I am then mistaking what does what. Sorry.

So this is my `flux-system/kustomization.yaml`. As you see, at the bottom it has variable substitution patched into the `environments/test/flux/flux-system/gotk-sync.yaml` file:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- gotk-sync.yaml
- gotk-components.yaml
patches:
  - target:
      kind: Service
      name: webhook-receiver
    patch: |
      - op: add
        path: "/metadata/annotations/cloud.google.com~1backend-config"
        value: '{"default": "webhook-receiver"}'
  - target:
      version: v1
      group: apps
      kind: Deployment
      name: image-reflector-controller
      namespace: flux-system
    patch: |-
      - op: add
        path: /spec/template/spec/containers/0/args/-
        value: --gcp-autologin-for-gcr
  - target:
      version: v1beta2
      group: kustomize.toolkit.fluxcd.io
      kind: Kustomization
      name: flux-system
      namespace: flux-system
    patch: |-
      - op: add
        path: /spec/postBuild
        value:
          substitute:
            environment: "test"
            ip_address: "..."
            project_id: "..."
            shared_project_id: "..."
...
```

I use that to pass the variables down the line to other Kostomizations. I am really confused about how this works at that point. One example would be this cert-manager kustomization.

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: cert-manager
  namespace: flux-system
spec:
  prune: True
  interval: 1m
  path: environments/${environment}/infrastructure/cert-manager
  dependsOn:
    - name: namespaces
  sourceRef:
    kind: GitRepository
    name: flux-system
    namespace: flux-system
  postBuild:
    substitute:
      environment: ${environment}
      ip_address: ${ip_address}
```

**Comment by kingdonb** [comment]
The `flux-system` kustomization has this limitation, because of the way that bootstrapping Flux works it has to remain independent of the environment, else it can be broken trivially by an error in a substitution. It's really critical for certain edge use cases that Flux bootstrap is not interrupted, else the cluster in a distant location could be left with no way to recover it without an on-site visit.

The root of the environment can be placed in a downstream Kustomization, such as `apps` tenant from the flux guides.

You can also take advantage of tools like Kyverno or others that copy configmaps and secrets – if there is some environment data that you want to have replicated between environments, why not encode that in a policy directive, rather than injecting from flux-system?

It's presented this way to be compatible with multi-tenancy. So if you need a configmap or secret in a downstream tenant, it will have to be present in the namespace. You'll use `substituteFrom` in the downstream Kustomization, rather than a string-ified substitution that goes through `envsubst`. This is easier to statically verify IMHO as there's no way to tell where `${environment}` should come from at the injection site, except for to fire up a runtime environment and run `flux trace`.

Please consider restructuring the configs with this advice, then revert to us with your issues if this cannot work for some reason that I have not foreseen! Thank you for engaging the detail in this discussion, it is probably an FAQ, if we can figure out how to express it in terms that will be easily google-able.

**Comment by ar-qun** [comment]
Thank you for the detailed explanation. I am still confused but I am going forward with your suggestion with `substituteFrom `and, if I understood you correctly, not using environment substitution on `Kustomizations `that `gotk-sync.yaml` points at directly.

Just in case, I am leaving here the big picture of where things are actually happening maybe it will clarify the issue and give some addition insights.

```
.
├── README.MD
└─── environments
│   ├── dev
│   │   ├── application # a kustomization also points here
│   │   │   ├── resource001.yaml
│   │   │   └── resource002.yaml
│   │   ├── flux edited manually
│   │   │   ├── flux-system fully deployed with terraform and not edited manually
│   │   │   │   ├── gotk-components.yaml # untoched
│   │   │   │   ├── gotk-sync.yaml # untoched
│   │   │   │   └── kustomization.yaml # patching the files above with environment substitution added to the mix
│   │   │   └── flux_kustomization.yaml # all Flux's kustomizations are here, one customization has {environment} in the name
│   │   └── infrastructure Flux's # kustomizations point here at every directory
│   │       ├── istio/...
│   │       └── grafana/...
│   └── prod <- exactly same as dev
│       ├── application
│       ├── flux
|       └── infrastructure
...

```

 Thanks again!
